File: src/front/train-booking-frontend/src/services/booking.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../environments/environment';

export interface Booking {
  id: number;
  user: any;
  train: any;
  seatsBooked: number;
  bookingDate: string;
  status: string;
}

@Injectable({
  providedIn: 'root'
})
export class BookingService {
  private apiUrl = `${environment.apiUrl}/booking`;

  constructor(private http: HttpClient) {}

  bookTrain(trainId: number, seatsToBook: number): Observable<any> {
    return this.http.post(`${this.apiUrl}/book`, {
      trainId: trainId.toString(),
      seatsToBook: seatsToBook.toString()
    });
  }

  getMyBookings(): Observable<Booking[]> {
    return this.http.get<Booking[]>(`${this.apiUrl}/my-bookings`);
  }

  getAllBookings(): Observable<Booking[]> {
    return this.http.get<Booking[]>(`${this.apiUrl}/all`);
  }

  getBookingById(id: number): Observable<Booking> {
    return this.http.get<Booking>(`${this.apiUrl}/${id}`);
  }

  cancelBooking(id: number): Observable<any> {
    return this.http.put(`${this.apiUrl}/cancel/${id}`, {});
  }

  getBookingsByTrainId(trainId: number): Observable<Booking[]> {
    return this.http.get<Booking[]>(`${this.apiUrl}/train/${trainId}`);
  }
}

File: src/main/java/project/service/BookingService.java
package project.service;

import org.springframework.http.ResponseEntity;
import project.entity.Booking;

import java.util.List;
import java.util.Map;

public interface BookingService {


    ResponseEntity<String> createBooking(Map<String, String> requestMap);


    ResponseEntity<List<Booking>> getMyBookings();


    ResponseEntity<List<Booking>> getAllBookings();


    ResponseEntity<Booking> getBookingById(Integer id);


    ResponseEntity<String> cancelBooking(Integer id);


    ResponseEntity<List<Booking>> getBookingsByTrainId(Integer trainId);
}